### ISSUE TITLE
Using sightglass to benchmark shootout

### ISSUE THREAD

**Issue by jlb6740** [fastly/lucet#102]
Hi ... What are the steps for building and running shootout with sightglass? Also, can you elaborate on the description that says that sightglass it is for testing "various versions" of the Lucet toolchain?

> Sightglass ships with a set of microbenchmarks called shootout. The scripts to build the shootout tests with native and various versions of the Lucet toolchain are in /benchmarks/shootout.

Some context: As far as I understand the docker container contains one version of Lucet. I've been able to run and time my own simple fib example that I copied to the container. Sightglass and Lucet readmes though talk about comparing multiple shared libraries and versions of the Lucet toolchain. What does that goal mean in this context of having one Lucet instance running?

Also when trying to build Sightglass in docker I got this error:

```
root@yshou-mint:/lucet/sightglass# cargo run
   Compiling proc-macro2 v0.4.27
   Compiling num-traits v0.2.6
   Compiling unicode-xid v0.1.0
   Compiling semver-parser v0.7.0
   Compiling cc v1.0.31
   Compiling libc v0.2.50
   Compiling syn v0.15.29
   Compiling num-integer v0.1.39
error: couldn't create a temp dir: No such file or directory (os error 2) at path "/tmp/rustcVkLQ6h"
error: aborting due to previous error
error: Could not compile `unicode-xid`.
warning: build failed, waiting for other jobs to finish...
```

Is sightglass meant to run in docker?

Also separately, I noticed a benchmark folder on host with a shootout directory and a makefile. I could not get the make command to build, but is this intended to be used if Lucet is running in docker and not on host? How does this folder relate to the base sightglass folder which also contains a shootout directory with the actual kernels included?

Thanks

**Comment by jedisct1** [comment]
Hi,

Sightglass compares different implementations of the same functions.

It works by loading shared libraries containing overlapping sets of functions, and will report how the function named `fx` in library `ly` compares to `fx` defined in library `ly'`. 

It's not specifically designed for Lucet, and works at least on Linux, macOS and BSD. You can use it to compare `gcc` vs `clang`, or to compare different optimization levels for a given platform.

But in this container, we produce two shared libraries with the same benchmarks: one using `clang` compiling directly to native code, and another one compiling to WebAssembly before using Lucetc to produce the native code.

Sightglass has its own repository: https://github.com/fastly/sightglass which includes some updates over the version currently shipped with Lucet. Maybe this is a better place to discuss Sightglass-specific issues.

It can run in the Docker container. The error you are seeing may be caused by the filesystem allocated to Docker running out of space.

However, if you installed Lucet (using `. devenv_setenv.sh`), you already have Sightglass installed. The command is simply called `sightglass` in the host, and in the Docker guest, it will also be in your `PATH` after sourcing `/opt/lucet/bin/devenv_setenv.sh`.

If you compile it independently, using release builds is highly recommended (`cargo build --release` or `cargo install ...`). The overhead of non-optimized code in debug builds may be significant compared to what you are actually trying to measure.

**Comment by jedisct1** [comment]
Also, in order to run the Lucet benchmark, type:

```sh
cd /lucet
make bench
```

in the container.

Alternatively, type `./devenv_run.sh make bench` in the host.

Compilation with `clang` can take quite some time due to the `switch` benchmark.

**Comment by jedisct1** [comment]
By the day, we will probably publish a blog post on Sightglass soon.

That will be a good opportunity to also improve its documentation.

**Comment by jlb6740** [comment]
Ok thanks a lot. So this is a way to directly compare C->WASM->Native with C->Native. This is nice. 

So I did a make bench but ended up with an build error:

error: Found argument '--reserved-size' which wasn't expected, or isn't valid in this context
        Did you mean --min-reserved-size?

I submitted a patch but see it has already been patched. I also saw an error:

No such file or directory (os error 2)
Makefile:37: recipe for target 'run' failed
make[1]: *** [run] Error 1
make[1]: Leaving directory '/lucet/benchmarks/shootout'
Makefile:32: recipe for target 'bench' failed
make: *** [bench] Error 2

Looks like the make bench command actually runs as well (I see many shootout results), but the above error looks unexpected.
